Update App Banner deep links for iOS universal links

frosty opened the pull request:
**Disclaimer:** I'm not a web developer, and I think there may be better ways to do this! Please let me know if something is wrong or can be improved.

This PR updates the mobile app banner deep link URLs for iOS.

![simulator-screen-shot-iphone-8-2018-07-04-at-17-22-06](https://user-images.githubusercontent.com/4780/42832134-2ededfe8-89e8-11e8-8f38-c2dff3c3ac11.png)

The app banners now point to `https://apps.wordpress.com/get`, which will redirect the user to the App Store if they don't have the WordPress for iOS app installed.

If the user does have the app installed, the app will be launched and we'll attempt to navigate them into the appropriate section of the app. To accomplish this, we also append a URL fragment to the base URL for each route.

* **Editor:** `https://apps.wordpress.com/get#post`
* **Notifications:** `https://apps.wordpress.com/get#notifications`
* **Reader:** `https://apps.wordpress.com/get#read` for the root of the Reader. If the user is in a specific section of the Reader, we'll attempt to pass through the entire feed URL so we can load the same one in the app. For example, `https://apps.wordpress.com/get#read/feeds/123456/posts/123456`
* **Stats:** As with the Reader, we'll attempt to append the current path as the fragment, so we can show the same section in the app. For example, `https://apps.wordpress.com/get#stats/discover.wordpress.com`

### To test

Unfortunately the deep linking can't easily be fully tested unless you have a development build of the iOS app. I may be able to put together a special alpha build if you really want to test this, but I think it's sufficient to verify that the URLs we're constructing are correct (I've already tested the launching of the app myself with a development build).

* If you don't have the WordPress app for iOS installed, or you don't have a development version, tapping the **Open in App** button on each of the 4 app banners should launch the App Store (via apps.wordpress.com). The App Store should show the WordPress app listing.
* Please verify that the URLs for each of the **Open in App** buttons match the schemas described above.

dmsnell commented on client/blocks/app-banner/index.jsx:
```diff
@@ -131,9 +132,21 @@ class AppBanner extends Component {
+				case NOTES:
+					return 'https://apps.wordpress.com/get#notifications';
+				case READER:
+					if ( currentPathFragment.length === 0 ) {
```
why are we special-casing the Reader with the fragment length check?

frosty replied:
The Reader doesn't have its own path when you view it on the web – so the root of the Reader just sits at `WordPress.com`. Each of the other routes have their own distinct path – `/stats`, `/notifications`, `/post`. The Reader can also be accessed at `/read`, but this handles the case where you're just at the root and there's no path, and ensures the app knows which section to take the user to.

dmsnell replied:
hm. it's still a bit unclear how this is supposed to function. if Reader is a special exception then I think it begs for a strong explanatory comment otherwise someone will come here and trash the code at some point in the future unless it's obvious why it's there :wink:

frosty replied:
Definitely a comment makes sense. I'll try and explain here a bit more clearly too with a couple of examples:

* If you visit the editor in a mobile web browser, the URL is `wordpress.com/post`, and you'll see the editor app banner. If the app is launched from here, we handle the path `/post` and display the editor.
* If you visit Discover in the Reader in a mobile web browser, the URL is `wordpress.com/discover` and you'll see the Reader app banner. If the app is launched from here, we handle the path `/discover`, and display the Discover feed in the app's Reader.
* If you visit the Reader (tap the Reader icon in the top bar) in a mobile web browser, the URL is just `wordpress.com`, and you'll see the Reader app banner. If the app is launched from here, without a special case, we'd just get the path `/`, and have no way to differentiate it from the logged out WordPress.com URL (which we don't want to deep link into the app). Hence, a special case to interpret this as `/read` in the app.

gwwar replied:
A one line comment should be sufficient. We basically want WordPress.com root or `/read` in this case?

frosty replied:
Essentially – I've added an explanatory comment to the code.

dmsnell commented on client/blocks/app-banner/index.jsx:
```diff
@@ -131,9 +132,21 @@ class AppBanner extends Component {
+					return 'https://apps.wordpress.com/get#' + currentPathFragment;
+				case STATS:
+					return 'https://apps.wordpress.com/get#' + currentPathFragment;
+			}
```
a more stable way of dealing with the fragment could be to encode the actual URL in its own transport and let the app disambiguate. I think at least we need some clarification in an explanatory comment _why_ we're passing the fragment and path like this.

for instance, imagine a fully-controlled URL…

```js
const routing = encodeURIComponent( JSON.stringify( {
	section: 'post',
	previousFragment: currentPathFragment,
} );
return `https://apps.wordpress.com/get?routing=${ routing }`
```

and we can make a helper for it…

```js
const urlFor = section => `https://apps.wordpress.com/get?${ encodeURIComponent( JSON.stringify( {
	section,
	previousFragment: currentPathFragment,
} ) ) }`;

case EDITOR:
	return urlFor( 'post' );

case NOTES:
	return urlFor( 'notifications' );

case READER:
	return urlFor( 'read' );
```

something like that

frosty replied:
I can certainly test that. The main benefit of passing the path in the way I am currently is that on the app side of things I can just pass the path directly to our routing mechanism for deep linked URLs coming from WordPress.com. It wouldn't be too much extra work to parse the necessary parts out of this type of URL though.

dmsnell replied:
I'm mainly concerned about bugs due to encoding issues in the URL

we're assuming a lot if we push out the `currentPathFragment` directly into the URL and especially if we're using a fragment as well 🤷‍♂️ 

just want to make sure that we at least recognize the places this will tend to break and determine how to or how much to mitigate them

frosty replied:
Thanks for raising this. I've used `encodeURIComponent` on the fragment now – please let me know if I need to do more than this.

dmsnell commented on client/blocks/app-banner/utils.js:
```diff
@@ -61,6 +63,13 @@ export function getCurrentSection( currentSection, isNotesOpen ) {
+export function getCurrentPathFragment( state ) {
+	const route = getCurrentRoute( state );
+
+	// Strip off the leading forward slash for use as a fragment
```
an example of the input data could make this comment much more useful to an unfamiliar reader - what does `getCurrentRoute()` return?

```js
// strip off the leading forward-slash for use as a fragment
// e.g. /settings/mysite.wordpress.com -> settings/mysite.wordpress.com
```

also, why are we stripping the slash?

frosty replied:
Definitely agree with the clearer example. If I'm completely honest, there's reason for stripping the leading slash other than I thought it looked nicer. I understand that's not a great reason, and perhaps it'd be best to just leave it in there :)

gwwar replied:
Seems like a bit of overkill to add this trivial util function before we have a second instance, it's easy enough to strip a leading slash in the deep link function as well.

gwwar replied:
This is technically a selector too, which would normally live in `state/selectors` I don't think we need this though. It's enough to modify in `mapStateToProps` or in the deep link function directly.

gwwar commented on client/blocks/app-banner/index.jsx:
```diff
@@ -131,9 +132,21 @@ class AppBanner extends Component {
+						return 'https://apps.wordpress.com/get#read';
+					}
+
+					return 'https://apps.wordpress.com/get#' + currentPathFragment;
```
Optional, but [temperate literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) are a bit easier to read than string concat.

eg:
```js
return `https://apps.wordpress.com/get#${ currentPathFragment }`;
```

frosty replied:
Thanks! With my outdated JavaScript knowledge, I didn't know this was a thing :) Definitely looks nicer.

gwwar commented on client/blocks/app-banner/index.jsx:
```diff
@@ -131,9 +132,21 @@ class AppBanner extends Component {
-		//TODO: update when deferred deep links are available
 		if ( this.isiOS() ) {
-			return 'itms://itunes.apple.com/us/app/wordpress/id335703880?mt=8';
+			switch ( currentSection ) {
```
If we want to verify that we return the correct app links, consider adding a unit test. It's fine to pull this switch out into it's own function (makes it easier to test).

```js
//...
if ( this.isiOS() ) {
    return getIOSDeepLinks( currentSection ); 
}
//...
}

function getIOSDeepLinks( currentSection ) {
    switch ( currentSection ) {
      //...
}
```
And in the test file:
```
//...
	test( 'iOS Deep links return correct URIs for READER', () => {
        expect( '/' ).toEqual( https://apps.wordpress.com/get#read );
        expect( '/read/feeds/12345/posts/6789' ).toEqual( 'https://apps.wordpress.com/get#read/feeds/12345/posts/6789' );
	} );
    test( 'iOS Deep links return correct URIs for STATS', () => {
        //...
	} );
//...
```

frosty replied:
Thanks – I've pulled out a function and added some tests!

gwwar commented on client/blocks/app-banner/index.jsx:
```diff
@@ -201,6 +214,7 @@ const mapStateToProps = state => {
 	return {
 		dismissedUntil: getPreference( state, APP_BANNER_DISMISS_TIMES_PREFERENCE ),
 		currentSection: getCurrentSection( sectionName, isNotesOpen ),
+		currentPathFragment: getCurrentPathFragment( state ),
```
Consider passing through  `getCurrentRoute( state )` instead, and letting the getDeepLink function(s) strip out the leading slash. It's much easier to read with that context, instead of needing to click through and see what a current path fragment is.

```
const route = getCurrentRoute( state ); //this may be null

frosty replied:
Thanks, I've removed this dependency and added some null checking.

gwwar commented:
Thanks @frosty ! I left a few notes, but looks pretty close, ping me if you need help with anything.

> Unfortunately the deep linking can't easily be fully tested unless you have a development build of the iOS app.

When is app update planned to be released? After this PR gets approval, we'll want to mark when this is safe to land.

gwwar commented:
For folks 👀 Here are more detailed testing instructions:

### Testing instructions
1. Use your physical Mobile device using this calypso.live branch: https://calypso.live/?branch=update/app-banner-ios-urls
2. Visit the reader, stats, editor, and notifications
3. Banners should be visible and behave as expected.
4. Open notifications again. Click on any part of the banner should not toggle notifications. Open in app and dismiss banner should work as expected.

In Desktop Chrome:
1. This banner should not be visible on any page in non-mobile browsers.
2. In order to test it, you can open up DevTools in Chrome and use `Toggle device toolbar`. Along making the screen size smaller, this will also change the `userAgent` which we use to test whether the app banner should be shown. Another option would be to select the `userAgent` manually in Chrome DevTools by following option 2 of [this guide](https://www.technipages.com/google-chrome-change-user-agent-string).
3. Verify that app banner is shown on stats, reader, and editor pages. It should also be shown on any page when notification panel is open.
4. Verify that correct text and icons are shown for each page as described in https://github.com/Automattic/wp-calypso/issues/15228.
5. Verify that impression tracks event has been recorded upon banner rendering.
6. Verify that `Open in app` redirects to app store and that appropriate tracks event/stat bump is recorded. (will be broken since the developer build won't be out).
7. Verify that `No thanks` closes the banner and that appropriate tracks event/stat bump is recorded.

**_Note:_** If you've dismissed the widget and want to bring it back for testing in current session, you can run the following line in your console: 
```javascript
dispatch( { type: 'PREFERENCES_SET', key: 'appBannerDismissTimes', value: null } );
```

gwwar commented:
Some lights smoke testing in emulation mode looks good 👍 
![Uploading Screen Shot 2018-07-17 at 11.28.31 AM.png…]()
<img width="753" alt="screen shot 2018-07-17 at 11 28 26 am" src="https://user-images.githubusercontent.com/1270189/42838145-d92307ba-89b4-11e8-93af-07c9fd43aee6.png">

sirreal commented on client/blocks/app-banner/test/app-banner.jsx:
```diff
@@ -0,0 +1,61 @@
+/**
+ * External dependencies
+ */
+import { expect } from 'chai';
```
Let's avoid introducing new usage of `chai`.

- Remove this import line.
- Jest will provide the global `expect` –no additional action required
- `expect( … ).equal( 'some-string' )` becomes `expect( … ).toBe( 'some-string' )`

frosty replied:
Thanks – I've pushed a fix for this.

frosty commented:
@dmsnell @gwwar I've pushed some updates, and I think I've addressed your comments – but please let me know if there's anything else.

> When is app update planned to be released? After this PR gets approval, we'll want to mark when this is safe to land.

So I expect this to go into version 10.6 of WordPress for iOS, which will go into testing on July 30th. However I think it would be okay to land this as soon as it's ready, as the links should still just redirect to the App Store before that app update is available and installed. Once the app enters testing, we'd have a 2 week period to verify / fix anything that may crop up. But I'm happy to delay it if you'd prefer though.

dmsnell commented on client/blocks/app-banner/index.jsx:
```diff
@@ -131,14 +131,38 @@ class AppBanner extends Component {
+				break;
+		}
+
+		return `https://apps.wordpress.com/get#${ encodeURIComponent( fragment ) }`;
```
note that we've exposed a case where we are using an unset variable - if `currentSection` isn't one of these four things then we'll be trying to encode `undefined`

frosty replied:
Thanks for catching that – I'm still in Swift mode where switches must be exhaustive and you'll get an error if they're not.

vindl replied:
With current code the `default` case will never be reached anyway, since `AppBanner` will bail early in render if one of those sections is not detected, see: https://github.com/Automattic/wp-calypso/blob/82f7c5fbd6905bc99a9638ea9075d7db48a14092/client/blocks/app-banner/index.jsx#L149 Still, that's not immediately obvious and it doesn't hurt to add safeguards here too. :+1:

dmsnell commented on client/blocks/app-banner/test/app-banner.jsx:
```diff
@@ -0,0 +1,72 @@
+		expect( wrapper.instance().getiOSDeepLink( null, STATS ) ).toBe(
+			'https://apps.wordpress.com/get#%2Fstats'
+		);
+	} );
```
Thanks for writing these tests. Don't let this be a blocker for the PR but I don't particularly find them that useful - they are mainly just making sure that somebody didn't change the constants defined in the code.

A  more valuable test might focus more on the behaviors - probably a good point to question the tests if the descriptions are like "should do the right thing"

So what are the behaviors here we want to test? What could unexpectedly go wrong?

```js
test( 'properly encodes tricky fragments', () => {
	expect( wrapper.instance().getiOSDeepLink( '?://&#', STATS ).split( '#' )[ 0 ] ).toEqual( '%3F%3A%2F%2F%26%23' );
} );
```

also, I know there was some discussion about methods vs. functions and I don't want to confuse you, but as you rightly questioned, testing is harder as methods. there's _no_ dependence on the instance here so let's move it out of the class and make it a function in the module scope that the class calls. the tests become more straight-forward

```js
import { getiOSDeepLink } from 'blocks/app-banner';

test( 'properly encodes tricky fragments', () => {
	expect( getiOSDeepLink( '?://&#', STATS ).split( '#' )[ 0 ] ).toEqual( '%3F%3A%2F%2F%26%23' );
} );
```

dmsnell commented on client/blocks/app-banner/index.jsx:
```diff
@@ -218,6 +193,45 @@ export class AppBanner extends Component {
+		return baseURI;
+	}
+
+	return `${ baseURI }#${ fragment }`;
```
fun fact! we can merge these into a single expression

```js
return ( fragment.length > 0 )
	? `${ baseURI }#${ fragment }`
	: baseURI;
```

if you like it that is

dmsnell commented on client/blocks/app-banner/index.jsx:
```diff
@@ -218,6 +193,45 @@ export class AppBanner extends Component {
+			break;
+	}
+
+	return encodeURIComponent( fragment );
```
just in case you are interested there are two ways we can remove the mutation here

we can return directly…

```js
const hasRoute = …

switch ( currentSection ) {
	case EDITOR:
		return encodeURIComponent( '/post' );

	case NOTES:
		return encodeURIComponent( '/notifications' );

	…

	default:
		return '';
}
```

we can also separate and compose the functionality in various ways. since this is pretty specific behavior and the function doesn't do much else, I'm going to create a function inside of it. could legitimately create another separate and testable function

```js
export function buildDeepLinkFragment( currentRoute, currentSection ) {
	const hasRoute = …;

	const getFragment = () => {
		switch ( currentSection ) {
			case EDITOR:
				return '/post';

			case NOTES:
				return '/notifications';

			case READER:
				return hasRoute ? currentRoute : '/read';

			case STATS:
				return hasRoute ? currentRoute : '/stats';

			default:
				return '';
		}
	}

	return encodeURIComponent( getFragment() );
}
```

this is mainly a style thing, just for inspiration

frosty replied:
Nice – I like the second option here (I wanted to avoid duplicating `encodeURIComponent` for each return).

frosty commented:
@dmsnell Do you know why the canary tests are failing? It doesn't look to me like anything I've done?

dmsnell commented on client/blocks/app-banner/test/app-banner.jsx:
```diff
@@ -0,0 +1,63 @@
+
+describe( 'iOS deep link fragments', () => {
+	test( 'properly encodes tricky fragments', () => {
+		expect( buildDeepLinkFragment( '?://&#', STATS ) ).toEqual( '%3F%3A%2F%2F%26%23' );
```
even though I suggested writing this code, I wonder how much this ends up conflating the tests here. we might consider using this instead…

```js
const funnyFragment = '?://&#';

expect( buildDeepLinkFragment( funnyFragment, STATS ) ).toEqual( encodeURIComponent( funnyFragment ) );
```

that way the test isn't dependent on the implementation of `encodeURIComponent()` (not that it will change any time soon) and it also indicates in the test code the intended behavior more semantically - we get the expected intention instead of the expected outcome

Copons commented:
Poor @frosty, he came in for a quick fix, and instead got a Full Fledged @dmsnell Lesson™ on `this` in JS. 😢
